Add tictactoe project files

bmwant opened the pull request:
**Where to put new files:**

- New files should go into a top-level subfolder, named after the article slug. For example: `my-awesome-article`

**How to merge your changes:** 

1. [Make sure the CI code style tests all pass (+ run the automatic code formatter if necessary).](https://github.com/realpython/materials/blob/master/README.md)
2. Find an RP Team member on Slack and ask them to review & approve your PR.
3. Once the PR has one positive ("approved") review, GitHub lets you merge the PR.
4. 🎉

PR for the article itself is [here](https://github.com/realpython/tutorial-drafts/pull/601)

bmwant commented:
@somacdivad could you review this?

somacdivad commented on tictactoe/io.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class ConsoleFrontend(IOFrontend):
+    placeholders = ["①", "②", "③", "④", "⑤", "⑥", "⑦", "⑧", "⑨"]
```
I'm concerned that we don't have a basic frontend that will work on everyone's terminal. Not all terminals will display emojis. I think it would be really nice to have a simple frontend that just worked for everyone, then a fancier frontend (like the existing `ConsoleFrontend`) to illustrate how to plug in different frontends.

Here's what I'm thinking as a super simple frontend:

```python
class BasicConsoleFrontend(IOFrontend):
    def print_board(self, board):
        for i, row in enumerate(board):
            for j, column in enumerate(row):
                if cell != Cell.EMPTY:
                    print(cell, end = "|")
                else:
                    cell_number = i * len(row) + j + 1
                    print(cell_number)
            print()
        print()

    def print_winner(self, name=None):
        if name is None:
            print("Game is a draw!")
        else:
            print(f"Player {name} wins!")

    def get_input(self):
        return input("Enter a number of the cell: ")
```

You could also simplify the nested loop a bit:

```python
for i, row in enumerate(board):
    row_display = []
    for j, col in enumerate(row):
        if cell != Cell.EMPTY:
            row_display.append[cell]
        else:
            cell_number = i * len(row) + j + 1
            row_display.append[cell_number]
    print(" | ".join(row_display)
```

That changes the output a bit so that there isn't the | on the far right-hand side of the table, and avoids the two calls to `print()` without any arguments.

So lots of food for thought there. Curious to know what you think!

bmwant replied:
I believe every modern terminal/shell should support proper unicode. Anyway agree to simplify this to just ASCII. The point of doing that initially was a comment from @gahjelle to have some nicer UI.

bmwant replied:
![Screen Shot 2020-08-25 at 10 57 12 AM](https://user-images.githubusercontent.com/3602533/91148039-c1a34200-e6c1-11ea-88f7-49e55b955863.png)

somacdivad replied:
@gahjelle Does love his pretty terminal UIs 🙂

> I believe every modern terminal/shell should support proper Unicode.

Not all readers use "modern" terminals, though. I think having ASCII as a default is the best option. It still looks good and is pretty much guaranteed to work anywhere.

somacdivad commented on tictactoe/io.py:
```diff
@@ -0,0 +1,75 @@
+        for i, row in enumerate(board):
+            table_row = []
+            for j, column in enumerate(row):
+                if (cell := board[i][j]) == Cell.X:
+                    text = click.style(cell.value, fg="red", bold=True)
+                elif cell == Cell.O:
+                    text = click.style(cell.value, fg="blue", bold=True)
+                else:
+                    text = str(i * len(row) + j + 1)
```
Here's another spot where we could simplify things a little bit:

```python
cell = board[i][j]
if cell != Cell.EMPTY:
    color = "red" if cell == Cell.X else "blue"
    text = click.style(cell.value, fg=color, bold=True)
else:
    text = str(i * len(row) + j + 1)
```

It might even be a good idea to refactor some of this into it's own function, but I'll let you decide on that...

bmwant commented on tictactoe/__init__.py:
```diff
@@ -0,0 +1,7 @@
+from enum import Enum
```
yes, but `player.py` also needs `Cell.EMPTY` check, so to prevent circular import we need it to be stored in a "shared" place or having an import within a `get_turn` method which I think would be a bad practice. I might also move this to separate `cell.py` module in case `__init__.py` is not the right place.

somacdivad replied:
Yeah, I missed that it was also used in `player.py` (left the comment before I had read that file)

bmwant commented on tictactoe/game.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class Game(object):
+    def __init__(
+        self, x_player=None, o_player=None, frontend: IOFrontend = None
```
Agree that type hint here seems like redundant

bmwant commented on tictactoe/game.py:
```diff
@@ -0,0 +1,90 @@
+    def _check_winner(self) -> Optional[Cell]:
+        # Check rows
+        for row in self.board:
+            if len(set(row)) == 1 and row[0] != Cell.EMPTY:
```
I've changed code to have separate `_check_winning_set` with shared logic, but I think it's a bit better to have explicit blocks highlighting which `group` we are checking at the moment even with some extra code duplication.

bmwant commented on tictactoe/game.py:
```diff
@@ -0,0 +1,90 @@
+        return self._check_draw()
+
+    def make_turn(self, turn: int, piece: Cell):
+        size = len(self.board)
+        i = turn // size
+        j = turn % size
```
created `Turn` class to deal with that

somacdivad commented on tictactoe/player.py:
```diff
@@ -2,18 +2,25 @@
 
     @abstractmethod
-    def get_turn(self, board) -> int:
+    def get_turn(self, board) -> Turn:
```
Since we agreed to remove type hints, I thinkwe can remove this, too:

```suggestion
    def get_turn(self, board):
```

somacdivad commented on tictactoe/player.py:
```diff
@@ -25,21 +32,24 @@ def __init__(self):
         super().__init__(name=name, frontend=frontend)
 
-    def get_turn(self, board) -> int:
+    def get_turn(self, board) -> Turn:
```
Here's another type hint that snuck through 🙂

```suggestion
    def get_turn(self, board):
```

bmwant commented:
hey @somacdivad, I've moved all the code to subfolder to be able to invoke it via `python -m`. Initially you supposed to run it from parent directory and this way is described in the article but here I think it's ok to have inner folder. Also I have added a `requirements.txt` file and instructions to the `README.md` file (creation of virtual environment is also described in the article itself).
